utils/src/main/java/software/amazon/awssdk/utils/async/DelegatingBufferingSubscriber.java

```diff
@@ -0,0 +1,137 @@
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * This Subscriber buffers the {@link ByteBuffer} it receives until the specififed maxiimumBuffer is reached, at which point
+ * it sends those bytes to the delegate {@link Subscriber}.
+ */
+@SdkInternalApi
+public class DelegatingBufferingSubscriber extends DelegatingSubscriber<ByteBuffer, ByteBuffer> {
+
+    /**
+     * The maximum amount of bytes allowed to be stored in the StoringSubscriber
+     */
+    private final int maximumBuffer;
+
+    /**
+     * Current amount of bytes buffered in the StoringSubscriber
+     */
+    private final AtomicLong currentlyBuffered = new AtomicLong(0);
+
+    /**
+     * Stores the bytes received from the upstream publisher, awaiting sending them to the delegate once the buffer size is
+     * reached.
+     */
+    private final StoringSubscriber<ByteBuffer> storage = new StoringSubscriber<>(Integer.MAX_VALUE);
+
+    /**
+     * Publisher to send the ByteBuffer received from upstream publisher to the StoringSubscriber
+     */
+    private final SimplePublisher<ByteBuffer> publisherToStorage = new SimplePublisher<>();
+
+    /**
+     * Subscription of the publisher this subscriber subscribe to
+     */
+    private Subscription subscription;
+
+    public DelegatingBufferingSubscriber(int maximumBuffer, Subscriber<? super ByteBuffer> subscriber) {
+        super(subscriber);
+        this.maximumBuffer = maximumBuffer;
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        super.onSubscribe(subscription);
+        this.subscription = subscription;
+        publisherToStorage.subscribe(storage);
+        subscription.request(maximumBuffer);
+    }
+
+    @Override
+    public void onNext(ByteBuffer byteBuffer) {
+        // First, if the incoming ByteBuffer would make the total buffered amount exceed the maximum buffer, send what is
+        // currently buffered in the storage to the delegate.
+        if (currentlyBuffered.get() + byteBuffer.remaining() > maximumBuffer) {
+            flushStorageToDelegate();
+        }
+
+        // Then, if the incoming ByteBuffer would still alone bust the buffer size, remove one chunk of it (of maximumBuffer
+        // size) then try again
+        if (currentlyBuffered.get() + byteBuffer.remaining() > maximumBuffer) {
+            ByteBuffer chunk = ByteBuffer.allocate(maximumBuffer);
+            while (chunk.hasRemaining()) {
+                chunk.put(byteBuffer.get());
+            }
+            chunk.position(0);
+            sendBytesToStorage(chunk);
+            onNext(byteBuffer);
+            return;
+        }
+
+        sendBytesToStorage(byteBuffer);
+        if (currentlyBuffered.get() >= maximumBuffer) {
+            flushStorageToDelegate();
+        }
```

Comment:
When would this be needed? Aren't the previous checks sufficient, or could there be extra data added on somehow?